tests/std/tests/P1004R2_constexpr_vector_bool/test.cpp

```diff
@@ -16,6 +16,39 @@ using namespace std;
 static constexpr bool input[]         = {true, false, true, true, false, true};
 static constexpr bool input_flipped[] = {false, true, false, false, true, false};
 
+struct demoterator { // demote pointer to input iterator
+    using iterator_category = std::input_iterator_tag;
+    using value_type        = int;
+    using difference_type   = std::ptrdiff_t;
+    using reference         = const int&;
+    using pointer           = void;

```

Comment:
Since this test is C++20-only, we could get by defining only `value_type` and `difference_type` here. Don't bother to change if we don't need to touch something else.

Reply:
I am not sure whether we could make such a change. I see N4901 [algorithms.requirements]/4.1 "If an algorithm’s template parameter is named `InputIterator`, `InputIterator1`, or `InputIterator2`, the template argument shall meet the *Cpp17InputIterator* requirements (23.3.5.3)." Then [vector.bool] specifies `template<class InputIterator> constexpr iterator insert(const_iterator position, InputIterator first, InputIterator last);`.

I don't see anything saying that `vector<bool>::insert()` can accept C++20 input iterators. (I also can't find the paragraph that makes `InputIterator` in the Containers clause have the same Word Of Power meaning as in the Algorithms clause, although I thought such a paragraph existed.)

Reply:
[algorithms.requirements] only applies to function templates that appear in subclauses of [algorithms]. The requirement that `vector<bool>::insert` takes Cpp17InputIterators comes from many miles away in [[sequence.reqmts]/3.5](https://eel.is/c++draft/sequence.reqmts#3.5). More importantly, removing all nested types other than `value_type` and `difference_type` would not alter the fact that this type meets the Cpp17InputIterator requirements; C++20's `iterator_traits` is smart enough to fill in the rest of the types if you declare `value_type` and `difference_type` ([[iterator.traits]/3.2](https://eel.is/c++draft/iterator.traits#3.2)).

Reply:
[tab:inputiterator] says that *Cpp17InputIterator* requires `*r++` to be well-formed and "convertible to `T`", but `demoterator::operator++(int)` returns `void`.

Reply:
That requirement is broken enough (LWG-484, LWG2962) that I tend to just ignore it.

Reply:
Ah, ok. No other changes were necessary, so I'm going to resolve this without changes.